### ISSUE TITLE
Protocol Handler: SVG does not render when used in img HTML tag

### ISSUE THREAD

**Issue by lidel** [mozilla/libdweb#54]
### Summary

SVG does not render in HTML when loaded via `<img src=`, even when loaded from the same origin.

### How to Reproduce

1. Run `demo:protocol` from my `svg-render-bug-demo` branch:
    ```
    git clone -b svg-render-bug-demo --depth 1 https://github.com/lidel/libdweb.git
    cd libdweb
    yarn && yarn demo:protocol
    ```
1. Open `dweb://html/dot.svg` directly – it [renders just fine](https://user-images.githubusercontent.com/157609/43783975-51921132-9a63-11e8-9d21-5a2c21dc772c.png). Test SVG comes from [commons](https://commons.wikimedia.org/wiki/File:ByColor_gwDot.svg).
1. Open `dweb://html/`, you will see inline `<svg>` rendering ok but external one loaded via `<img src="dweb://html/dot.svg" />` (same origin) failing to load with [errors in console](https://user-images.githubusercontent.com/157609/43784025-6eb7e106-9a63-11e8-8ca6-c14f57d45852.png)
   - If I try to open SVG directly via [right click](https://user-images.githubusercontent.com/157609/43784227-decc41bc-9a63-11e8-99e5-a4d2c83bd689.png) it loads fine again.
1. Open `dweb://foo/`, you will see `<img src="dweb://html/dot.svg" />` (loaded from different origin) fails to load just like in previous step
   - What is different, if I try to open SVG directly via [right click](https://user-images.githubusercontent.com/157609/43784227-decc41bc-9a63-11e8-99e5-a4d2c83bd689.png) it does not work and [produces additional error in console](https://user-images.githubusercontent.com/157609/43784024-6e6da9ce-9a63-11e8-911c-62d665ff57f8.png). Is this the same problem space as  #52, or should I create a separate issue for it? 

### Screenshot

> ![screenshot](https://user-images.githubusercontent.com/157609/43784786-37c17a52-9a65-11e8-8f41-ccbbe99e39c6.png)

**Comment by Gozala** [comment]
>4. Open dweb://foo/, you will see <img src="dweb://html/dot.svg" /> (loaded from different origin) fails to load just like in previous step
> 
>    -  What is different, if I try to open SVG directly via right click it does not work and produces additional error in console. Is this the same problem space as #52, or should I create a separate issue for it?

No it's different from #52 as in this case they are under same protocol but different origins, created #55 to track that.

**Comment by Gozala** [comment]
Here is what I see observing the output produced:

- Request goes across to a handler
  ```json
  { 
     "type":"request",
     "requestID":"dweb:2:Agent@2#58998",
     "url":"dweb://html/dot.svg",
     "scheme":"dweb"
  }
  ```
- Head of the response makes it across to the content process
   ```json
    {"type":"head","requestID":"dweb:2:Agent@2#58998", "head":{}}
    ```
- Request listener is started `listener.onStartRequest(this, context)` with default content metadata:
  ```json
  {
    "contentType": "application/x-unknown-content-type",
    "contentLength": -1,
    "contentCharset": "utf-8"
  }
  ```

- Then response Body makes it over to the content process and `nsIArrayBufferInputStream` is created.
- Now what's **strange part** is that [`cancel`](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIRequest#cancel()) with code `2152988677`  is called on the corresponding channel / request and that happens before previous call is does `listener.onDataAvailable`, which makes me wonder if something is spawning a nested event loop there.
- Then the `NS_BINDING_ABORTED` is forwarded to the protocol handler
- Protocol handler responds with `{"type": "end"}` message
- End message makes it's way to content process channel but attempt to `loadGroup.removeRequest` throws exception.
- Only after all this it seems that `listener.onDataAvailable` is called but it throws exception `NS_BINDING_ABORTED`, presumably because `listener.onStopRequest` got invoked by `End` message that somehow managed to happen before.

I'll continue digging through this to better determine what is going on.

**Comment by Gozala** [comment]
Looking further into this it appears that maybe call to `listener.onDataAvailable` triggers a `cancel()` call.

**Comment by lidel** [comment]
Thanks for looking into this.
IIUC, basically if we solve #56  then this should be a non-issue?

**Comment by Gozala** [comment]
> Thanks for looking into this.
> IIUC, basically if we solve #56 then this should be a non-issue?

I have resolved #56 which should provide a way to address #56 as you pointed out. That being said I'd still keep this open as I would like to:

1. Find out how same html with embedded image would've behaved if contentType was not set by the server. Ideally protocol behavior should match exactly that.
2. Find out if we can get same behavior as with `file:///` protocol.

**Comment by Gozala** [comment]
I did preliminary digging into `file:` protocol implementation and it seems to be handled very differently, it seems to use [`nsIFileChannel`](https://dxr.mozilla.org/mozilla-central/source/netwerk/protocol/file/nsIFileChannel.idl) which does not extend [`nsIChannel`](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIChannel) so I presume that code path for `file:` is very different and not applicable to this use case.

**Comment by Gozala** [comment]
It looks like if headers are omitted in HTTP server firefox correctly sniffs the content type, here is the code I used for testing:

```js
const http = require('http')

const server = http.createServer((request, response) => {
  response.writeHead(200)
  if (request.url.endsWith('dot.svg')) {
    response.end(`<?xml version="1.0" encoding="UTF-8"?>
    <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
    <svg version="1.1" xmlns="http://www.w3.org/2000/svg" x="0px" y="0px" width="100px" height="150px" viewBox="0 0 10 15" enable-background="new 0 0 10 15" xml:space="preserve">
    <g>
      <path fill="#FFFFFF" d="M5,11c-1.929,0-3.5-1.571-3.5-3.5S3.071,4,5,4s3.5,1.571,3.5,3.5S6.929,11,5,11z"/>
      <path fill="#808080" d="M5,4.588c1.609,0,2.912,1.303,2.912,2.912S6.609,10.412,5,10.412S2.088,9.109,2.088,7.5 S3.391,4.588,5,4.588 M5,3C2.519,3,0.5,5.019,0.5,7.5S2.519,12,5,12s4.5-2.019,4.5-4.5S7.481,3,5,3L5,3z"/>
    </g>
    </svg>`)
  } else {
    response.end(`<h1>HTML</h1><p><img src="dot.svg" alt="dot.svg" border="1" style="width:100px;height:150px;" /></p>`)
  }
})

server.listen(8090)
```

Digging through the `nsIHttpChannel` implementation it appears that it first checks the `Content-Type` headers and if not present it uses sniffs through the content to figure out and set appropriate `contentType`.  Which is pretty much what we are proposing to do here.

**Comment by Gozala** [comment]
I end up landing #60 to which does more or less what `nsIHttpChannel` does. Which is if response does not contain `contentType` it will use [`nsIContentSniffer`](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIContentSniffer) from first data chunk as it arrives.

@lidel With this change your original example seems to work as expected. If I recall correctly you also mentioned to me in person that there was some issue with text file / or maybe markdown file mime type detection. I would expect this to be handled by this change as well.

**Comment by lidel** [comment]
Ack, #60 fixed rendering of SVGs:

> ![slrvlyd](https://user-images.githubusercontent.com/157609/44027391-a6a9e694-9ef6-11e8-85ea-223eeb4f4948.png)

Regarding Markdown: it opens fine if content type is set to `text/plain`. 
If set to `text/markdown` browser tries to download it instead of rendering, but the same is true for `http://`, so no issue here, I think.
